NMS-13362: Add Kafka consumer for OpenNMS events (#3495)
OpenNMS/opennms: docs/modules/operation/pages/events/sources/kafka.adoc

[[ga-events-sources-kafka]]

= Kafka Consumer (Events)

> Events can also be created by routing them to a specific topic on Kafka.

Reviewer: You can also create events by routing them to a specific topic on Kafka.

Reviewer: ```suggestion
You can also create events by routing them to a specific topic on Kafka.
```

> The Kafka Consumer feature allows events to be consumed from specific topic on Kafka.

Reviewer: The Kafka Consumer feature allows events to be consumed from a specific topic on Kafka.

Reviewer: ```suggestion
The Kafka Consumer feature allows events to be consumed from a specific topic on Kafka.
```

> Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]

Reviewer: Create events in protobuf fomrat link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)].

> Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]

Reviewer: Events should be created in protobuf format

Reviewer: ```suggestion
Create events in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)].
```

> See `kafka-consumer-events.proto`  for the model definitions.

Reviewer 2: ```suggestion
See `kafka-consumer-events.proto` for the model definitions.
```

> The Kafka Consumer by default consumes events from topic named `opennms-events` and forward them to `Eventd`.

Reviewer: The Kafka Consumer by default consumes events from topics named `opennms-events` and forwards them to `Eventd`.

Reviewer: ```suggestion
The Kafka Consumer by default consumes events from topics named `opennms-events` and forwards them to `Eventd`.
```

== Enable Kafka Consumer

The Kafka Consumer is disabled by default and can be enabled as follows.

First, log in to the Karaf shell of your {page-component-title} instance and configure the Kafka client settings to point to your Kafka broker.
See link:https://kafka.apache.org/27/documentation.html#consumerconfigs[Consumer Configs] for a complete list of available options.

> [source]

Reviewer 2: ```suggestion
[source, console]
```

----

> $ ssh -p 8101 admin@localhost

Reviewer 2: ```suggestion
ssh -p 8101 admin@localhost
```

Reviewer 2: We should split the source sections to indicate what is console and what is Karaf shell like this:

```
.Configure the Kafka features and the Minion location via the Karaf Shell
[source, console]
----
ssh -p 8201 admin@localhost
----

.Configure the Event consumer client
[source, karaf]
----
config:edit org.opennms.features.kafka.consumer.client
config:property-set bootstrap.servers 127.0.0.1:9092
config:update
----
```

...

> admin@opennms()> config:edit org.opennms.features.kafka.consumer.client

Reviewer 2: ```suggestion
config:edit org.opennms.features.kafka.consumer.client
```

> admin@opennms()> config:property-set bootstrap.servers 127.0.0.1:9092

Reviewer 2: ```suggestion
config:property-set bootstrap.servers 127.0.0.1:9092
```

Reviewer 2: We should probably add a callout for the 127.0.0.1 bootstrap server and port like this:

> Connect to the following Kafka nodes and adjust the IPs or FQDNs with the Kafka port (9092) accordingly.

> admin@opennms()> config:update

Reviewer 2: ```suggestion
config:update
```

----

Next, install the `opennms-kafka-consumer` feature from that same shell using:

[source]
----

> admin@opennms()> feature:install opennms-kafka-consumer

Reviewer 2: ```suggestion
feature:install opennms-kafka-consumer
```

----

To ensure that the feature continues to be installed on subsequent restarts, add `opennms-kafka-consumer` to the `featuresBoot` property in the `$\{OPENNMS_HOME}/etc/org.apache.karaf.features.cfg`.

== Configure Topic Name

.Karaf login, configuration of events topic

> [source]

Reviewer 2: Same as above, split and no prompts so we can use copy & paste.

config:update
----

> NOTE: While configuring events topic, make sure that it doesn't conflict with other topics in OpenNMS subsystem. If you are unsure, keep the default one.

Reviewer: NOTE: While configuring the eventsTopic, make sure that it doesn't conflict with other topics in the OpenNMS subsystem. If you are unsure, keep the default one.

Reviewer: ```suggestion
NOTE: While configuring the eventsTopic, make sure that it doesn't conflict with other topics in the OpenNMS subsystem. If you are unsure, keep the default one.
```
